topdown/eval.go

```diff
@@ -1175,10 +1175,16 @@ func (e *eval) saveCall(declArgsLen int, terms []*ast.Term, iter unifyIterator)
 
 func (e *eval) saveInlinedNegatedExprs(exprs []*ast.Expr, iter unifyIterator) error {
 
-	// This function does not include with statements on the exprs because
-	// they will have already been saved and therefore had their any relevant
-	// with statements set.
+	with := make([]*ast.With, len(e.query[e.index].With))
+
+	for i := range e.query[e.index].With {
+		cpy := e.query[e.index].With[i].Copy()
+		cpy.Value = e.bindings.PlugNamespaced(cpy.Value, e.caller.bindings)
+		with[i] = cpy
+	}
+
```

Comment:
If we are always doing it here does that mean we can remove doing it elsewhere? Or sometimes we'll just double apply the `with` statements? Or they were never added in the first place?

Reply:
The old comment was wrong--when the evaluator evaluates `not` it removes the `with` statements from complemented expression (this happens inside `evalNotPartial`). It does this because the `with` statements have already been processed. Since we may be saving the result of evaluating the complement (which had the `with` statements dropped) we need to add them back here. Similarly, since the `evalNotPartial` plugs the save stack expressions, we need to plug the with statements here.

Reply:
Thanks! That makes more sense now 👍